ext/intl/listformatter/listformatter_class.c

```diff
@@ -0,0 +1,240 @@
+        RETURN_THROWS();
+    }
+
+    if (locale_len > INTL_MAX_LOCALE_LEN) {
+        zend_argument_value_error(1, "Locale string too long, should be no longer than %d characters", INTL_MAX_LOCALE_LEN);
+        RETURN_THROWS();
+    }
+
+    UErrorCode status = U_ZERO_ERROR;
+    #if U_ICU_VERSION_MAJOR_NUM >= 67
+        if (type != ULISTFMT_TYPE_AND && type != ULISTFMT_TYPE_OR && type != ULISTFMT_TYPE_UNITS) {
+            zend_argument_value_error(2, "must be one of IntlListFormatter::TYPE_AND, IntlListFormatter::TYPE_OR, or IntlListFormatter::TYPE_UNITS");
+            RETURN_THROWS();
+        }
+        
+        if (width != ULISTFMT_WIDTH_WIDE && width != ULISTFMT_WIDTH_SHORT && width != ULISTFMT_WIDTH_NARROW) {
+            zend_argument_value_error(3, "must be one of IntlListFormatter::WIDTH_WIDE, IntlListFormatter::WIDTH_SHORT, or IntlListFormatter::WIDTH_NARROW");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_openForType(locale, type, width, &status);
+    #else
+        if (type != ULISTFMT_TYPE_AND) {
+            zend_argument_value_error(2, "ICU 66 and below only support IntlListFormatter::TYPE_AND");
+            RETURN_THROWS();
+        }
+
+        if (width != ULISTFMT_WIDTH_WIDE) {
+            zend_argument_value_error(3, "ICU 66 and below only support IntlListFormatter::WIDTH_WIDE");
+            RETURN_THROWS();
+        }
+
+        LISTFORMATTER_OBJECT(obj) = ulistfmt_open(locale, &status);
+    #endif
+
+    if (U_FAILURE(status)) {
+        intl_error_set(NULL, status, "Constructor failed", 0);
+        zend_throw_exception(IntlException_ce_ptr, "Constructor failed", 0);
+        RETURN_THROWS();
+    }
+}
+/* }}} */
+
+/* {{{ listformatter_format */
+PHP_METHOD(IntlListFormatter, format)
+{
+    ListFormatter_object *obj = Z_INTL_LISTFORMATTER_P(ZEND_THIS);
+    zval *strings;
+
+    ZEND_PARSE_PARAMETERS_START(1, 1)
+        Z_PARAM_ARRAY(strings)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (!LISTFORMATTER_OBJECT(obj)) {
+        intl_error_set(NULL, U_ILLEGAL_ARGUMENT_ERROR, "ListFormatter not properly constructed", 0);
+        RETURN_FALSE;
+    }
+
+    HashTable *ht = Z_ARRVAL_P(strings);
+    uint32_t count = zend_array_count(ht);
+    if (count == 0) {
+        RETURN_EMPTY_STRING();
+    }
+
+    const UChar **items = (const UChar **)safe_emalloc(count, sizeof(const UChar *), 0);
+    int32_t *itemLengths = (int32_t *)safe_emalloc(count, sizeof(int32_t), 0);
+    uint32_t i = 0;
+    zval *val;
+
+    ZEND_HASH_FOREACH_VAL(ht, val) {
+        zend_string *str_val;
+        
+        str_val = zval_get_string(val);
+        
+        // Convert PHP string to UTF-16
+        UChar *ustr = NULL;
+        int32_t ustr_len = 0;
+        UErrorCode status = U_ZERO_ERROR;
+        
+        intl_convert_utf8_to_utf16(&ustr, &ustr_len, ZSTR_VAL(str_val), ZSTR_LEN(str_val), &status);
+        if (U_FAILURE(status)) {
+            efree(items);
```

Comment:
in theory, items[index so and so] might have make it until this failure thus you re gonna get some leaks.

Reply:
Oh, that's right. Now it should be fixed.